CSharp/JobPrep/Program.cs

```diff
@@ -0,0 +1,333 @@
+                            Log("add vm", en.Current.Name);
+                            vms.Add(en.Current.Name);
+                            if (ct.IsCancellationRequested)
+                                break;
+                        }
+                    }));
+
+                    // launch N threads to get job prep task status of the nodes in the queue
+                    for (int i = 0; i < parallelism; i++)
+                        tasks.Add(Task.Run(() =>
+                        {
+                            while (!ct.IsCancellationRequested)
+                            {
+                                string vm;
+                                try {
+                                    vm = vms.Take(ct);
+                                } catch // if token cancelled, abort the thread
+                                {
+                                    break;
+                                }
+                                Log("start polling vm", vm);
+                                try
+                                {
+                                    var state = pm.GetJobPreparationTaskStatus(pool, vm, workitem, job);
+                                    Log("got vm state", vm);
+                                    if (state.State == JobPreparationTaskState.Completed) // job prep task completed on the thread
+                                    {
+                                        if (state.ExitCode == 0)
+                                            Interlocked.Increment(ref nb_succeeded);
+                                        else
+                                            Interlocked.Increment(ref nb_failed);
+                                        Log("==== job prep done ====", vm);
+                                        if (stop(nb_succeeded, nb_failed))
+                                            cts.Cancel();
+                                    }
+                                    else // job prep still running
+                                    {
+                                        Log("state not completed", vm);
+                                        // Sleep for some time and put the VM back to the queue
+                                        Task.Run(() => { Task.Delay(freq); if (!ct.IsCancellationRequested) { Log("add vm back", vm); vms.Add(vm); } });
+                                    }
+                                }
+                                catch // job prep task status not found - not run on this VM
+                                {
+                                    Log("error query", vm);
+                                    Task.Run(() => {Task.Delay(freq); if (!ct.IsCancellationRequested) { Log("add vm back", vm); vms.Add(vm); } });
+                                }
+                            }
+                            Log("cancel");
+                        }));
+                    Task.WaitAll(tasks.ToArray(), timeout);
```

Comment:
So, I think there may be timing bugs with the current setup.  You're putting your tasks to add VM names to your collection into the same array as the tasks to pull items from the collection, but you don't wait for all the "add tasks" to finish before running the "remove tasks".  Then, when a "remove task" fails to find an item, you use a break statement to exit your loop, and you effectively have 1 fewer "remove task" doing any work. Although your "add tasks" will complete quickly, I don't think you have a guarantee that you won't start removing VM names from your collection before you're done adding them.  I think it's technically possible that you'll add a VM name, then you'll have 5 "remove tasks" kick off.  One will do work, and the other 4 will exit immediately.  Then, you're other 2 VM names get added, and you don't have any tasks waiting for them.  Does that make sense?  I could be wrong here too, these kinds of things can be complicated :)